python/ray/serve/_private/common.py

```diff
@@ -118,17 +118,76 @@ class DeploymentStatusTrigger(str, Enum):
     DELETING = "DELETING"
 
 
+# Internal Enum used to manage deployment state machine
+class DeploymentStatusInternalTrigger(str, Enum):
+    HEALTHY = "HEALTHY"
+    CONFIG_UPDATE = "CONFIG_UPDATE"
+    AUTOSCALE_UP = "AUTOSCALE_UP"
+    AUTOSCALE_DOWN = "AUTOSCALE_DOWN"
+    MANUALLY_INCREASE_NUM_REPLICAS = "MANUALLY_INCREASE_NUM_REPLICAS"
+    MANUALLY_DECREASE_NUM_REPLICAS = "MANUALLY_DECREASE_NUM_REPLICAS"
+    REPLICA_STARTUP_FAILED = "REPLICA_STARTUP_FAILED"
+    HEALTH_CHECK_FAILED = "HEALTH_CHECK_FAILED"
+    INTERNAL_ERROR = "INTERNAL_ERROR"
+    DELETE = "DELETE"
+
+
 @dataclass(eq=True)
 class DeploymentStatusInfo:
     name: str
     status: DeploymentStatus
     status_trigger: DeploymentStatusTrigger
     message: str = ""
 
+    @staticmethod
+    def ranking_order() -> List[
+        Tuple[DeploymentStatus, Optional[DeploymentStatusTrigger]]
+    ]:
+        """List of states in ranked order.
+
+        Each ranked state has the format of a tuple with either 1 or 2 items.
+        If 1 item: contains a single DeploymentStatus, representing states with
+            that DeploymentStatus and any DeploymentStatusTrigger.
+        If 2 items: tuple contains a DeploymentStatus and a DeploymentStatusTrigger,
+            representing a state with that status and status trigger.
+        """
+
+        return [
+            # Status ranking order is defined in a following fashion:
+            #   1. (Highest) State signalling any failures in the system
+            (DeploymentStatus.UNHEALTHY,),
+            #   2. States signaling the user updated the configuration.
+            (DeploymentStatus.UPDATING,),
+            (DeploymentStatus.UPSCALING, DeploymentStatusTrigger.CONFIG_UPDATE_STARTED),
+            (
+                DeploymentStatus.DOWNSCALING,
+                DeploymentStatusTrigger.CONFIG_UPDATE_STARTED,
+            ),
+            #   3. Steady state or autoscaling.
+            (DeploymentStatus.UPSCALING, DeploymentStatusTrigger.AUTOSCALING),
+            (DeploymentStatus.DOWNSCALING, DeploymentStatusTrigger.AUTOSCALING),
+            (DeploymentStatus.HEALTHY,),
+        ]
+
+    @property
+    def rank(self) -> int:
+        """Get priority of state based on ranking_order().
+
+        The ranked order indicates what the status should be of a
+        hierarchically "higher" resource when derived from a group of
+        `DeploymentStatusInfo` sub-resources.
+        """
+
+        for i, s in enumerate(DeploymentStatusInfo.ranking_order()):
+            if len(s) == 1 and s[0] == self.status:
+                return i
+            elif len(s) == 2 and s == (self.status, self.status_trigger):
+                return i
+
     def debug_string(self):
         return json.dumps(asdict(self), indent=4)
 
-    def update(
+    def _update(
```

Comment:
nit: I find "update" confusing here because it may be returning a completely new status (or new trigger). Perhaps calling it "transition" is better?

Reply:
Makes sense. I'm a little wary of using "transition" though, since it doesn't actually handle any transition logic, and simply functions as an "update the keys of the dictionary" kind of function. I've changed it to `_updated_copy`, how does that sound?

Reply:
sgtm